### ISSUE TITLE
[IFC] support labels with implication between principals

### ISSUE THREAD

**Issue by tiziano88** [project-oak/oak#1364]
This proposal is about extending the IFC model and corresponding labels in order to support implication (`→`) in addition to the previously agreed upon conjunction (`∧`) and disjunction (`∨`) operations that are part of most IFC lattices in literature.

Intuitively, implication would behave similar to [material implication](https://en.wikipedia.org/wiki/Material_implication_(rule_of_inference)) in [propositional logic](https://en.wikipedia.org/wiki/Propositional_calculus).

Given atomic principals `A` and `B`, the combined principal `A→B` (read "A implies B") represents the authority of B "according to" (or "subject to") A.

In particular, the following "flows to" relation applies:

(A→B)⊑(C→D)⇔(A⊑C)∧(B⊑D)

By the way, this also sounds very similar to "authority projections" or "owned principals" in [FLAM](http://www.cs.cornell.edu/andru/papers/flam/flam-csf15.pdf), so it may well be that it is exactly the same concept after all, just with a different name, and defined on the label lattice instead of the authority lattice.

I think this also nicely captures some aspects of delegation and remote attestation:

If P_0 and P_1 represent two different versions of the platform (including hardware and software, e.g. enclave + Oak runtime code), such that P_1 is meant to replace P_0, we can model this transition by making all the principals implied by their respective platform, and also assuming that P_0⊑P_1 (i.e. data can flow from the old to the new version of the platform), but also P_1⋢P_0, obviously (i.e. data cannot go back to a previous version of the platform, in order to prevent downgrade attacks and exploit existing bugs in the old version).

In this model, almost all principals may be "implied", to represent the fact that really they are "guaranteed" by some other underlying principal (the platform, at a minimum).

Also for things like storage, it may be useful to represent principals as implied by the storage system itself, so that for instace the principal for user U_0 would be implied by the principal for the storage system S, producing the combined principal S→U_0, which represents the fact that data using it as a confidentiality label may only flow to a location with this label if the source label flows to S in the first place.

For integrity labels, implied principals may represent the fact that data has been processed by a node, and keep a record of the integrity of the original data, but implied by the last node in the chain. For instance, if there is data with integrity label A, which is then processed by nodes with integrity B and then C, the final integrity may be represented as C→B→A, i.e. the integrity is really that of the last node C, but if the recipient is willing to trust C, then it will be able to conclude that the intermediate data had integrity B→A, and again, assuming it trusts B, it will be able to conclude that the original data had integrity A. Note that if the recipient does not trust C, then it is not affected by incorrect labels or judgement, which is line with the desired properties of a "sound" IFC.

cc @aferr @anghelcovici @seefeldb

Some (sparse) references:

- [Material Implication in Orthomodular (and Boolean) Lattices](https://projecteuclid.org/download/pdf_1/euclid.ndjfl/1093883401)
- [Flow-Limited Authorization](http://www.cs.cornell.edu/andru/papers/flam/flam-csf15.pdf)
- [First-Order Logic for Flow-Limited Authorization](https://www.cs.cornell.edu/~ethan/papers/flafol.pdf)

**Comment by seefeldb** [comment]
How would integrity look like for a node C that receives data from both A
and B? (Or more complex: E receiving from B->A and D->C).

**Comment by aferr** [comment]
I also like the idea of moving "->" into the syntax of labels. (I think there were e-mail threads about this topic before, and I'm trying to dig them up in case there is something useful).

I think the analogous concept to "implies" in FLAM (original, I don't know the details of FLAFOL yet) would actually just be our choice of "speaksfor" or "flowsTo" (rather than projections or owned principals) depending on how we define "->" over labels. Here is why:
- In the propositional lattice "->" is the order relation.
- In the authority lattice over principals (/ labels) "speaksFor" is the order relation.
- In the information flow over information flow labels (/ principals) "flowsTo" is the order relation

With the way you wrote it, it's more like "flowsTo".
We could also consider making it more like "speaksFor". If we do this, the delegation analogy might work better because "speaksFor" was a concept that FLAM borrowed from authorization to capture delegation.

**Comment by aferr** [comment]
Also, I don't yet have an opinion on whether it's better for "->" to be like "flowsTo" or "speaksFor". I just wanted to point out that those are two options.

**Comment by aferr** [comment]
Oh, but also if we do this we should really flip the meaning of /\ and \/ so that this matches the usual propositional lattice (rather than FLAM) :)

**Comment by tiziano88** [comment]
> How would integrity look like for a node C that receives data from both A and B? (Or more complex: E receiving from B->A and D->C).

Intuitively, I would say the integrity label of such node would be at most C→(A∨B), where A and B may be atomic or compound principals. @seefeldb does this make sense to you?

**Comment by seefeldb** [comment]
If it is C→(A∨B), can it also say C→(A∨B∨F) without having read from F? That would be bad.

**Comment by tiziano88** [comment]
I agree that would be bad, and also it would be against classical logic from a notation point of view, so I think there should be a natural way of only allowing the sensible derivations.

Need to think more about what this means exactly, but I'm confident it will work out.

https://en.m.wikipedia.org/wiki/Deduction_theorem

**Comment by seefeldb** [comment]
is C→(A) <  C→(A∨B) <  C→(A∨B∨F)?

That's why i thought that it's maybe C→(A∧B), i.e. if you trust C then that implies both A and B.

**Comment by tiziano88** [comment]
I am not sure what you mean by "<", could you clarify? Intuitively C→(A) is "more trusted" (a stronger endorsement) than C→(A∨B) , does that make sense to you?

C→(A∧B) does not seem to follow logically, and also intuitively I don't see how something would get endorsed to the combined authority of A and B if it is made of pieces with authority of either A or B (never combining their endorsement power).

**Comment by seefeldb** [comment]
> I am not sure what you mean by "<", could you clarify? Intuitively C→(A) is "more trusted" (a stronger endorsement) than C→(A∨B) , does that make sense to you?

Yes, I meant which I meant more trusted. And I of course meant C→(A) > C→(A∨B) > C→(A∨B∨F). Sorry.

> C→(A∧B) does not seem to follow logically, and also intuitively I don't see how something would get endorsed to the combined authority of A and B if it is made of pieces with authority of either A or B (never combining their endorsement power).

Let's say A implies "from a real GPS sensor" and B implies "from a real camera" (let's assume all on one device for now), then C might write a geolocated photo that is known to have an actually measured GPS coordinate that was captured with the real camera.

How are you interpreting integrity? The nature of C might shift the meaning of A and B, but that's part of trusting C.

If I pass this through a simple filter D that slightly improves the contrast, should I never leave B on? Or is it just a judgement call whether changing the contrast still means "from a real camera" or not. Since we retain this information with the syntax notation, I'm reading this as D→C→B which I might trust in one context but not another.

Then the new integrity is D, and →C→B are just extra information, and what's important about that is that it can't be faked (D can't just invent C or B). Whether I treat D like B is a decision based on trusting D→C→. But maybe it would be more accurate to interpret that as "B means raw from camera" and "D→C→B means no material changes to the photo after capturing". Maybe that stretches the meaning of → too much? But as a use-case, this is something we'd want to be able to capture.

**Comment by seefeldb** [comment]
Hmm, but yeah, in order to make Arcs's "is derived from" work easily, ∨ seems correct.

I.e. it's really "is at most derived from A or B" and so making up C→(A∨B∨F) doesn't help. It's important that C→(A∨B) can't be changed to C→A, which would pass a more restrictive policy.

Ok, so ∨ is right, and I need to think about the implications for the GPS + real camera example. What I don't like is that D could just add "∨B" to an A, i.e. if instead of D→C→A it could claim D→C→(A∨B) because that's weaker, then it's just about trust in D and IFC can't strengthen that (or statically check compliance, etc.).

**Comment by tiziano88** [comment]
> What I don't like is that D could just add "∨B" to an A.

In logic, adding "∨B" to any other predicate is weakening, so I don't see how that is problematic. See https://en.wikipedia.org/wiki/Disjunction_introduction .

In IFC, data that was labelled with an integrity label A can strictly "go to more places" and "pass more checks" than data labelled with A∨B.

Do we 100% agree on this @seefeldb , or do you have a different intuition?

> Let's say A implies "from a real GPS sensor" and B implies "from a real camera" (let's assume all on one device for now), then C might write a geolocated photo that is known to have an actually measured GPS coordinate that was captured with the real camera.
>
> How are you interpreting integrity? The nature of C might shift the meaning of A and B, but that's part of trusting C.

This is exactly my point in which really you should be trusting C to make this judgement, because IFC does not know about cameras or GPS location; otherwise how would you propagate the information that the pixels come from the camera and the EXIF metadata comes from GPS, with a single label?

> If I pass this through a simple filter D that slightly improves the contrast, should I never leave B on? Or is it just a judgement call whether changing the contrast still means "from a real camera" or not. Since we retain this information with the syntax notation, I'm reading this as D→C→B which I might trust in one context but not another.
>
> Then the new integrity is D, and →C→B are just extra information, and what's important about that is that it can't be faked (D can't just invent C or B). Whether I treat D like B is a decision based on trusting D→C→. But maybe it would be more accurate to interpret that as "B means raw from camera" and "D→C→B means no material changes to the photo after capturing". Maybe that stretches the meaning of → too much? But as a use-case, this is something we'd want to be able to capture.

Yes we agree that "it can't be faked", this follows logically from the fact that A is essentially the same as A→⊥ , and the only thing A can do out of thin air is to weaken ⊥ even more (which obviously cannot be weakened any more, since ⊥∨A==⊥).

Of course the reverse is not allowed, i.e. A cannot just become A→B "for free", since that would really mean A→⊥∧B, which would require a "witness" for B, at least.

**Comment by aferr** [comment]
http://www.scs.stanford.edu/~dm/home/papers/stefan:dclabels.pdf
